extensions/amp-date-countdown/1.0/date-countdown.js

```diff
@@ -47,95 +50,104 @@ const TimeUnit = {
 };
 
 // Default prop values
-const DEFAULT_OFFSET_SECONDS = 0;
 const DEFAULT_LOCALE = 'en';
 const DEFAULT_WHEN_ENDED = 'stop';
 const DEFAULT_BIGGEST_UNIT = 'DAYS';
 
+/**
+ * @param {!JsonObject} data
+ * @return {string}
+ */
+const DEFAULT_RENDER = (data) =>
+  /** @type {string} */ (`${data['days']} ${data['dd']}, ` +
+    `${data['hours']} ${data['hh']}, ` +
+    `${data['minutes']} ${data['mm']}, ` +
+    `${data['seconds']} ${data['ss']}, `);
+
 /**
  * @param {!DateCountdownPropsDef} props
  * @return {PreactDef.Renderable}
  */
 export function DateCountdown({
-  endDate,
-  timeleftMs,
-  timestampMs,
-  timestampSeconds,
-  offsetSeconds = DEFAULT_OFFSET_SECONDS,
+  dateTime,
   whenEnded = DEFAULT_WHEN_ENDED,
   locale = DEFAULT_LOCALE,
   biggestUnit = DEFAULT_BIGGEST_UNIT,
-  render,
-  children,
+  render = DEFAULT_RENDER,
   ...rest
 }) {
   useResourcesNotify();
   const {playable} = useAmpContext();
-  const epoch = useMemo(
-    () =>
-      getEpoch(endDate, timeleftMs, timestampMs, timestampSeconds) +
-      offsetSeconds * MILLISECONDS_IN_SECOND,
-    [endDate, timeleftMs, timestampMs, timestampSeconds, offsetSeconds]
+
+  const epoch = getDate(dateTime);
+
+  // How much time is left in our countdown
+  const setTimeleft = useState(epoch - Date.now())[1];
+
+  // One time calculation of time labels in specified locale
+  const localeStrings = useMemo(
+    () => getLocaleWord(/** @type {string} */ (locale)),
+    [locale]
   );
-  const [timeleft, setTimeleft] = useState(epoch - Date.now());
-  const localeStrings = useMemo(() => getLocaleWord(locale), [locale]);
+
+  // Reference to DOM element to get access to correct window
   const rootRef = useRef(null);
 
+  // A reference to our 'data', 'data' is used to populate the template
+  // A reference is used so we can use the same object instance between
+  // successive re-renders
+  const dataRef = useRef(
+    getYDHMSFromMs(
+      epoch - Date.now() + DELAY,
+      /** @type {string} */ (biggestUnit)
+    )
+  );
+
+  // Put localeStrings into the data object
+  Object.assign(/** @type {!Object} */ (dataRef.current), localeStrings);
```

Comment:
Reference the diagram: https://docs.google.com/document/d/1INUhhNEyVXArtNCuAG-y9evhV_3QoNwRf0GmrGTfEDE/edit#bookmark=id.m0gs38nvpn2h

This part can either be computed with every render (as is the case now) or can be initialized each time `dataRef.current` is re-calculated (would add to box 1 and box 9).  Since the operation is relatively inexpensive, I figured it made sense to include in every render and slightly "simplify the code".  (this part always stays the same, whereas the rest of `dataRef.current` is constantly re-calculated)

Reply:
I agree with Dima here. Something isn't right in this construction. Can you include this as part of each assignment to `dataRef.current` instead?

Reply:
Right!  In the latest revision, I've factored into a function `getDataForTemplate` which is only used in the assignment to `data` via `useMemo` (this replaces `dataRef.current`).